### DISCUSSION TITLE
Troubleshooting LIBUSB_ERROR_BUSY

### DISCUSSION THREAD

**Discussion by andrewbaxter** [libusb/libusb#1553] [Q&A] [answered]
I've seen many posts about `LIBUSB_ERROR_BUSY` but they all basically end with people guessing at possible other services holding a device open.  Is there a way to do this without guesswork?  If not, could some tool be added to aid in debugging?

**Comment by tormodvolden** [comment, marked as answer]
On linux you may try lsof on the /dev/bus/usb/xxx/yyy node. If it is something transitory like a udev action it becomes more difficult. You can strace individual processes to check for access to the nodes. Otherwise kernel tracing tools would be needed. On Windows you can also see which processes have a device object opened, and you can trace driver stacks if you have an idea about which to check.

But generally the operating systems are not very helpful, process isolation serves to hide all details from other processes, so all you get is the "busy" error when you do a system call.

**Reply by andrewbaxter** [reply]
Thanks, that was a super comprehensive answer! Ah okay, so this is from the kernel, not libusb itself...

I did try `fuser`, but it returned no processes.  It turned out in my case it was in use by `pn533_usb`... which I guess means I'd have needed kernel tracing tools if I hadn't stumbled upon that in a small miracle.

**Comment by tormodvolden** [comment]
And it doesn't have to be another process opening it, it can be a kernel driver being busy with the device also I suppose.

**Comment by tormodvolden** [comment]
If you do `git grep LIBUSB_ERROR_BUSY` on the libusb git tree, you'll see that the linux backend typically translates system call error -EBUSY to LIBUSB_ERROR_BUSY but there are other cases where libusb returns that error for other reasons, some related to internal book-keeping of transfers.
